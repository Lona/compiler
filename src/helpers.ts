import { IFS } from 'buffs'
import { createModule, ModuleContext } from './logic/module'
import { Config, load } from './utils/config'
import { defaultReporter, Reporter } from './utils/reporter'

/**
 * Helpers passed to every plugins. They contain:
 *
 * - methods abstracting the file system
 * - a centralized log reporter
 * - the workspace's configuration, etc.
 */
export type Helpers = {
  fs: IFS
  reporter: Reporter
  config: Config
  module: ModuleContext
}

export default async (
  fs: IFS,
  workspacePath: string,
  options: {
    outputPath?: unknown
    reporter?: Reporter
  } = {}
): Promise<Helpers> => {
  const { reporter = defaultReporter } = options

  const helpers: Helpers = {
    fs,
    reporter,
    config: await load(fs, workspacePath),
    module: createModule(fs, workspacePath),
  }

  return helpers
}
