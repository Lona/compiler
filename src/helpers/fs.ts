import * as fs from 'fs'
import * as path from 'path'

export type FSWrapper = {
  readFile(filePath: string): Promise<string>
  writeFile(filePath: string, data: string): Promise<void>
  copyDir(dirPath: string, output?: string): Promise<void>
}

export const createFSWrapper = (
  workspacePath: string,
  _outputPath?: unknown
): FSWrapper => {
  const outputPath =
    typeof _outputPath === 'string'
      ? _outputPath
      : path.join(process.cwd(), 'lona-generated')
  const fsWrapper = {
    readFile(filePath: string) {
      return fs.promises.readFile(
        path.resolve(workspacePath, filePath),
        'utf-8'
      )
    },
    writeFile(filePath: string, data: string) {
      const resolvedPath = path.resolve(outputPath, filePath)
      fs.mkdirSync(path.dirname(resolvedPath), { recursive: true })
      return fs.promises.writeFile(resolvedPath, data, 'utf-8')
    },
    async copyDir(dirPath: string, output: string = '.') {
      const resolvedPath = path.resolve(workspacePath, dirPath)
      const files = await fs.promises.readdir(resolvedPath)

      await Promise.all(
        files.map(async x => {
          if (
            (await fs.promises.stat(path.join(resolvedPath, x))).isDirectory()
          ) {
            return
          }

          return fsWrapper.writeFile(
            path.join(output, x),
            await fsWrapper.readFile(path.join(dirPath, x))
          )
        })
      )
    },
  }

  return fsWrapper
}
