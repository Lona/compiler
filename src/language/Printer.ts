import { doc, Doc } from 'prettier'
import { Definition, Value } from './Parser'
import { StateDefinition, Token, PrintPattern } from './Lexer'

const {
  builders: { concat },
} = doc

export function formatToken(
  lexerDefinition: StateDefinition[],
  token: Token
): Doc {
  const rules = lexerDefinition.flatMap(state => state.rules)

  const rule = rules.find(rule => rule.name === token.type)

  if (!rule) {
    throw new Error(`Couldn't find rule for token: ${token.type}`)
  }

  function formatPrintPattern(printPattern: PrintPattern): Doc {
    switch (printPattern.type) {
      case 'literal': {
        return printPattern.value
      }
      case 'indexReference': {
        return token.groups[printPattern.value]
      }
      case 'sequence': {
        return concat(printPattern.value.map(formatPrintPattern))
      }
    }
  }

  return rule.print ? formatPrintPattern(rule.print) : ''
}

export function formatTokens(
  lexerDefinition: StateDefinition[],
  tokens: Token[]
): Doc {
  return concat(tokens.map(formatToken.bind(null, lexerDefinition)))
}

export function print(
  document: Doc,
  options: doc.printer.Options = {
    printWidth: 80,
    tabWidth: 2,
    useTabs: false,
  }
) {
  return doc.printer.printDocToString(document, options).formatted
}
