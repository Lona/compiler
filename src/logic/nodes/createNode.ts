import { LogicAST as AST } from '@lona/serialization'
import { EnumerationDeclaration } from './EnumerationDeclaration'
import { FunctionCallExpression } from './FunctionCallExpression'
import { FunctionDeclaration } from './FunctionDeclaration'
import { IdentifierExpression } from './IdentifierExpression'
import {
  IDeclaration,
  IEvaluationContributor,
  IExpression,
  ILiteral,
  INode,
  IScopeContributor,
  ITypeCheckerContributor,
  ITypeAnnotation,
} from './interfaces'
import { LiteralExpression } from './LiteralExpression'
import {
  ArrayLiteral,
  BooleanLiteral,
  ColorLiteral,
  NoneLiteral,
  NumberLiteral,
  StringLiteral,
} from './literals'
import { MemberExpression } from './MemberExpression'
import { NamespaceDeclaration } from './NamespaceDeclaration'
import { RecordDeclaration } from './RecordDeclaration'
import { VariableDeclaration } from './VariableDeclaration'
import {
  IdentifierTypeAnnotation,
  FunctionTypeAnnotation,
} from './typeAnnotations'
import {
  FunctionParameter,
  FunctionParameterDefaultValue,
} from './FunctionParameter'

export function createTypeAnnotationNode(
  syntaxNode: AST.SyntaxNode
): ITypeAnnotation | undefined {
  switch (syntaxNode.type) {
    case 'typeIdentifier':
      return new IdentifierTypeAnnotation(syntaxNode)
    case 'functionType':
      return new FunctionTypeAnnotation(syntaxNode)
    default:
      return undefined
  }
}

export function createLiteralNode(
  syntaxNode: AST.SyntaxNode
): ILiteral | undefined {
  switch (syntaxNode.type) {
    case 'boolean':
      return new BooleanLiteral(syntaxNode)
    case 'number':
      return new NumberLiteral(syntaxNode)
    case 'string':
      return new StringLiteral(syntaxNode)
    case 'none':
      return new NoneLiteral(syntaxNode)
    case 'color':
      return new ColorLiteral(syntaxNode)
    case 'array':
      return new ArrayLiteral(syntaxNode)
    default:
      return undefined
  }
}

export function createExpressionNode(
  syntaxNode: AST.SyntaxNode
): IExpression | undefined {
  switch (syntaxNode.type) {
    case 'identifierExpression':
      return new IdentifierExpression(syntaxNode)
    case 'memberExpression':
      return new MemberExpression(syntaxNode)
    case 'functionCallExpression':
      return new FunctionCallExpression(syntaxNode)
    case 'literalExpression':
      return new LiteralExpression(syntaxNode)
    default:
      return undefined
  }
}

export function createDeclarationNode(
  syntaxNode: AST.SyntaxNode
): IDeclaration | undefined {
  switch (syntaxNode.type) {
    case 'variable':
      return new VariableDeclaration(syntaxNode)
    case 'record':
      return new RecordDeclaration(syntaxNode)
    case 'enumeration':
      return new EnumerationDeclaration(syntaxNode)
    case 'function':
      return new FunctionDeclaration(syntaxNode)
    case 'namespace':
      return new NamespaceDeclaration(syntaxNode)
    default:
      return undefined
  }
}

function isScopeVisitor(node: INode): node is IScopeContributor {
  return 'scopeEnter' in node || 'scopeLeave' in node
}

function isTypeCheckerVisitor(node: INode): node is ITypeCheckerContributor {
  return 'typeCheckerEnter' in node || 'typeCheckerLeave' in node
}

function isEvaluationVisitor(node: INode): node is IEvaluationContributor {
  return 'evaluationEnter' in node
}

const nodeCache: { [key: string]: INode } = {}

export function createNode(syntaxNode: AST.SyntaxNode): INode | undefined {
  const id = syntaxNode.data.id

  if (id in nodeCache) {
    return nodeCache[id]
  }

  let node: INode | undefined =
    createDeclarationNode(syntaxNode) ||
    createTypeAnnotationNode(syntaxNode) ||
    createExpressionNode(syntaxNode) ||
    createLiteralNode(syntaxNode)

  if (!node) {
    switch (syntaxNode.type) {
      case 'parameter':
        node = new FunctionParameter(syntaxNode as AST.FunctionParameter)
        break
      case 'value':
      case 'none':
        node = new FunctionParameterDefaultValue(
          syntaxNode as AST.FunctionParameterDefaultValue
        )
        break
    }
  }

  if (node) {
    nodeCache[id] = node
  }

  return node
}

export function createScopeVisitor(
  syntaxNode: AST.SyntaxNode
): IScopeContributor | undefined {
  const node = createNode(syntaxNode)

  return node && isScopeVisitor(node) ? node : undefined
}

export function createTypeCheckerVisitor(
  syntaxNode: AST.SyntaxNode
): ITypeCheckerContributor | undefined {
  const node = createNode(syntaxNode)

  return node && isTypeCheckerVisitor(node) ? node : undefined
}

export function createEvaluationVisitor(
  syntaxNode: AST.SyntaxNode
): IEvaluationContributor | undefined {
  const node = createNode(syntaxNode)

  return node && isEvaluationVisitor(node) ? node : undefined
}
