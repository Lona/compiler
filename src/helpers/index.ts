import { LogicAST } from '@lona/serialization'
import { config as Config } from '../utils'
import { generate as generateEvaluationContext } from './evaluation-context'
import { EvaluationContext } from './logic-evaluate'
import { createStandardLibraryResolver } from './hardcoded-mapping'
import { Reporter, defaultReporter } from './reporter'
import { FSWrapper, createFSWrapper } from './fs'
export { HardcodedMap } from './hardcoded-mapping'
export { EvaluationContext } from './logic-evaluate'
import { TraversalConfig, reduce } from './logic-traversal'
import {
  makeProgram,
  declarationPathTo,
  findParentNode,
  findNode,
} from './logic-ast'

/**
 * Helpers passed to every plugins. They contains some methods abstracting
 * the file system, the evaluation context of the workspace,
 * a reporter (to centralize the logs), the workspace's configuration, etc.
 */
export type Helpers = {
  fs: FSWrapper
  reporter: Reporter
  config: Config.Config
  /**
   * The evaluation context of the Lona Workspace.
   *
   * It contains some mapping between the identifiers and what they reference,
   * eg. `Optional.none` -> the function declaration in Prelude.logic
   * which is used to evaluate `Optional.none` to
   * `{ type: unit, memory: { type: 'unit' } }` for example. It takes care of
   * the scope of the identifier to determine the reference.
   *
   * Additionally, it keeps track of where declaration are from.
   */
  evaluationContext: EvaluationContext | undefined
  /**
   * Takes a hardcoded map `standard library -> node` and creates a function
   * which returns a node if the Logic node passed as argument is from
   * the standard library.
   *
   * Depending on the context, the hardcoded map might use the
   * evaluation context to evaluate the Logic node instead of trying to translate
   * the standard library method into the target format.
   */
  createStandardLibraryResolver: typeof createStandardLibraryResolver
  ast: {
    /** Tries to make a program out of any Logic node */
    makeProgram: (node: LogicAST.SyntaxNode) => LogicAST.Program | undefined
    /** A set of methods to help traverse the AST */
    traversal: {
      declarationPathTo: (
        node: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) => string[]
      /** Find the node with the matching id */
      findNode: (
        rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) =>
        | LogicAST.SyntaxNode
        | LogicAST.Pattern
        | LogicAST.Identifier
        | undefined
      /** Find the parent of the node with the matching id */
      findParentNode: (
        rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) => LogicAST.SyntaxNode | undefined
      /**
       * The `reduce()` method executes a reducer function (that you provide)
       * on each node of the AST, resulting in a single output value.
       *
       * Your reducer function's returned value is assigned to the accumulator,
       * whose value is remembered across each iteration throughout the AST,
       * and ultimately becomes the final, single resulting value.
       *
       * The traversal is depth-first, you can specify whether it should be
       * pre-order or post-order in the config.
       */
      reduce: <T>(
        node: LogicAST.SyntaxNode,
        callbackfn: (
          previousValue: T,
          currentNode: LogicAST.SyntaxNode,
          config: TraversalConfig
        ) => T,
        initialResult: T,
        config?: TraversalConfig
      ) => T
    }
  }
}

export type PreludeFlags = {
  [id: string]:
    | LogicAST.RecordDeclaration
    | LogicAST.EnumerationDeclaration
    | LogicAST.NamespaceDeclaration
}

export default async (
  workspacePath: string,
  outputPath?: unknown,
  _reporter?: Reporter
): Promise<Helpers> => {
  const fsWrapper = createFSWrapper(workspacePath, outputPath)

  const config = await Config.load(workspacePath, {
    forEvaluation: true,
    fs: fsWrapper,
  })

  let cachedEvaluationContext: EvaluationContext | undefined

  const reporter = _reporter || defaultReporter

  return {
    fs: fsWrapper,
    config,
    get evaluationContext() {
      if (cachedEvaluationContext) {
        return cachedEvaluationContext
      }
      cachedEvaluationContext = generateEvaluationContext(config, reporter)
      return cachedEvaluationContext
    },
    createStandardLibraryResolver,
    reporter,
    ast: {
      makeProgram,
      traversal: {
        declarationPathTo,
        findNode,
        findParentNode,
        reduce,
      },
    },
  }
}
