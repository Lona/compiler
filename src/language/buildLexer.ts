import { EnumerationDeclaration } from '../logic/nodes/EnumerationDeclaration'
import { Helpers } from '../helpers'
import { Lexer, Rule, Token, StateDefinition, Action } from './Lexer'
import { IdentifierExpression } from '../logic/nodes/IdentifierExpression'
import { LiteralExpression } from '../logic/nodes/LiteralExpression'
import { StringLiteral } from '../logic/nodes/literals'
import { LogicAST } from '@lona/serialization'
import { FunctionCallExpression } from '../logic/nodes/FunctionCallExpression'
import { isNode } from '../logic/ast'
import { valueBindingName } from '../plugins/swift/convert/codable'
import { IExpression } from '../logic/nodes/interfaces'

const getStringLiteral = (node: IExpression): string | undefined =>
  node instanceof LiteralExpression && node.literal instanceof StringLiteral
    ? node.literal.value
    : undefined

export function isTokenizer(node: EnumerationDeclaration): boolean {
  return node.attributes.some(
    attribute =>
      attribute.callee instanceof IdentifierExpression &&
      attribute.callee.name === 'tokenize'
  )
}

type TokenGenerator = {
  parentState: string
  rule: Rule
  transform: (token: Token) => unknown
}

function getAction(node: IExpression): Action {
  if (
    node instanceof FunctionCallExpression &&
    node.callee instanceof IdentifierExpression
  ) {
    switch (node.callee.name) {
      case 'pop':
        return { type: node.callee.name }
      case 'push':
      case 'next': {
        const value = getStringLiteral(node.argumentExpressionNodes[0])

        if (value) {
          return { type: node.callee.name, value }
        }
      }
    }
  }

  throw new Error('Invalid action')
}

export function getTokenAttributes(
  node: Extract<LogicAST.EnumerationCase, { type: 'enumerationCase' }>
): TokenGenerator[] {
  return node.data.attributes
    .map(attribute => new FunctionCallExpression(attribute))
    .flatMap(attribute => {
      if (
        attribute.callee instanceof IdentifierExpression &&
        attribute.callee.name === 'token'
      ) {
        const { pattern, state, action } = attribute.argumentExpressionNodes

        const rule = {
          name: node.data.name.name,
          pattern: getStringLiteral(pattern) ?? node.data.name.name,
          ...(action && { action: getAction(action) }),
        }

        const parentState = getStringLiteral(state) ?? 'main'

        const associatedValues = node.data.associatedValues.filter(isNode)

        return {
          parentState,
          rule,
          transform: token => {
            return {
              type: token.type,
              ...Object.fromEntries(
                associatedValues.map((associatedValue, index, array) => [
                  valueBindingName(associatedValue, index, array.length),
                  token.groups[index],
                ])
              ),
            }
          },
        }
      }
      return []
    })
}

export function buildLexer(
  node: EnumerationDeclaration,
  context: Helpers
): (source: string) => unknown {
  const generators = node.cases.flatMap(getTokenAttributes)

  const states: StateDefinition[] = []

  generators.forEach(generator => {
    const state = states.find(state => state.name === generator.parentState)

    if (state) {
      state.rules.push(generator.rule)
    } else {
      states.push({ name: generator.parentState, rules: [generator.rule] })
    }
  })

  const lexer = new Lexer(states, 'main')

  return source => {
    const tokens = lexer.tokenize(source)

    return tokens.map(token => {
      const generator = generators.find(
        generator => generator.rule.name === token.type
      )!
      return generator.transform(token)
    })
  }
}
