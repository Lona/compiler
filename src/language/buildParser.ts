import { Helpers } from '../helpers'
import { LogicAST as AST } from '@lona/serialization'
import { EnumerationDeclaration } from '../logic/nodes/EnumerationDeclaration'
import { FunctionCallExpression } from '../logic/nodes/FunctionCallExpression'
import { IdentifierExpression } from '../logic/nodes/IdentifierExpression'
import { IExpression } from '../logic/nodes/interfaces'
import { LiteralExpression } from '../logic/nodes/LiteralExpression'
import { ArrayLiteral } from '../logic/nodes/literals'
import { MemberExpression } from '../logic/nodes/MemberExpression'
import { RecordDeclaration } from '../logic/nodes/RecordDeclaration'
import { VariableDeclaration } from '../logic/nodes/VariableDeclaration'
import {
  Definition,
  Field,
  Parser,
  Pattern,
  RecordNodeDefinition,
  EnumNodeDefinition,
  OrPattern,
  field,
  manyField,
} from './Parser'
import {
  getPrintAttributeExpression,
  getNodePrintPattern,
  getFieldPrintPattern,
} from './buildPrinter'
import {
  FieldPrintPattern,
  sequencePrintPattern,
  joinCommandPrintPattern,
  tokenReferencePrintPattern,
  nodeReferencePrintPattern,
  selfReferencePrintPattern,
} from './Printer'
import { inspect } from 'util'

export function isParser(
  node: RecordDeclaration | EnumerationDeclaration
): boolean {
  return node.attributes.some(
    attribute =>
      attribute.callee instanceof IdentifierExpression &&
      attribute.callee.name === 'parse'
  )
}

function getPattern(resolvedSelf: string, node: IExpression): Pattern {
  if (
    node instanceof FunctionCallExpression &&
    node.callee instanceof IdentifierExpression
  ) {
    switch (node.callee.name) {
      case 'or':
        return {
          type: node.callee.name,
          value: Object.values(node.argumentExpressionNodes).map(
            getPattern.bind(null, resolvedSelf)
          ),
        }
      case 'many':
      case 'option': {
        return {
          type: node.callee.name,
          value: Object.values(node.argumentExpressionNodes).map(
            getPattern.bind(null, resolvedSelf)
          )[0],
        }
      }
    }
  } else if (
    node instanceof LiteralExpression &&
    node.literal instanceof ArrayLiteral
  ) {
    return {
      type: 'sequence',
      value: node.literal.elements.map(getPattern.bind(null, resolvedSelf)),
    }
  } else if (node instanceof MemberExpression) {
    const names = node.names.map(name =>
      name === 'self' ? resolvedSelf : name
    )

    if (names[0] === 'Token') {
      return {
        type: 'reference',
        value: {
          type: 'token',
          name: names[1],
        },
      }
    } else {
      return {
        type: 'reference',
        value: {
          type: 'field',
          nodeName: names[0],
          fieldName: names[1],
        },
      }
    }
  } else if (node instanceof IdentifierExpression) {
    return {
      type: 'reference',
      value: {
        type: 'node',
        name: node.name,
      },
    }
  }

  throw new Error('Invalid pattern')
}

export function getParseAttribute(
  name: string,
  attributes: FunctionCallExpression[]
): Pattern | undefined {
  const attribute = attributes.find(
    attribute =>
      attribute.callee instanceof IdentifierExpression &&
      attribute.callee.name === 'parse'
  )

  if (!attribute) return

  const { pattern } = attribute.argumentExpressionNodes

  return getPattern(name, pattern)
}

export function getEnumParseAttribute(
  node: EnumerationDeclaration
): Pattern | undefined {
  const attribute = node.attributes.find(
    attribute =>
      attribute.callee instanceof IdentifierExpression &&
      attribute.callee.name === 'parse'
  )

  if (!attribute) return

  const { pattern } = attribute.argumentExpressionNodes

  // Use default enum pattern
  if (!pattern) {
    return {
      type: 'or',
      value: node.cases.map(enumCase => {
        return {
          type: 'reference',
          value: {
            type: 'field',
            nodeName: node.name,
            fieldName: enumCase.data.name.name,
          },
        }
      }),
    }
  } else {
    return getParseAttribute(node.name, node.attributes)
  }
}

function inferFieldPattern(
  fieldName: string,
  annotation: Extract<AST.TypeAnnotation, { type: 'typeIdentifier' }>,
  nodeNames: string[]
): Pattern | undefined {
  const typeName = annotation.data.identifier.string

  if (nodeNames.includes(typeName)) {
    return {
      type: 'reference',
      value: { type: 'node', name: typeName },
    }
  } else if (
    typeName === 'Array' &&
    annotation.data.genericArguments[0].type === 'typeIdentifier'
  ) {
    const pattern = inferFieldPattern(
      fieldName,
      annotation.data.genericArguments[0],
      nodeNames
    )

    if (!pattern) return

    return {
      type: 'many',
      value: pattern,
    }
  } else if (
    typeName === 'Optional' &&
    annotation.data.genericArguments[0].type === 'typeIdentifier'
  ) {
    const pattern = inferFieldPattern(
      fieldName,
      annotation.data.genericArguments[0],
      nodeNames
    )

    if (!pattern) return

    return {
      type: 'option',
      value: pattern,
    }
  } else if (typeName === 'String') {
    return {
      type: 'reference',
      value: {
        type: 'token',
        name: fieldName,
      },
    }
  }
}

function getRecordNode(
  declaration: RecordDeclaration,
  nodeNames: string[]
): RecordNodeDefinition {
  const printAttribute = getPrintAttributeExpression(declaration.attributes)

  return {
    type: 'record',
    name: declaration.name,
    pattern: getParseAttribute(declaration.name, declaration.attributes)!,
    print: printAttribute ? getNodePrintPattern(printAttribute) : undefined,
    fields: declaration.variables.flatMap((variable): Field[] => {
      const pattern = getParseAttribute(variable.name, variable.attributes)
      const printAttribute = getPrintAttributeExpression(variable.attributes)

      if (pattern) {
        return [
          field({
            name: variable.name,
            pattern,
            ...(printAttribute && {
              print: getFieldPrintPattern(printAttribute),
            }),
          }),
        ]
      }

      const annotation = variable.syntaxNode.data.annotation

      // Try to infer a pattern from the type annotation
      if (annotation && annotation.type === 'typeIdentifier') {
        const pattern = inferFieldPattern(variable.name, annotation, nodeNames)

        if (pattern) {
          if (pattern.type === 'many') {
            return [
              manyField({
                name: variable.name,
                pattern,
                ...(printAttribute && {
                  print: getFieldPrintPattern(printAttribute),
                }),
              }),
            ]
          }

          return [
            field({
              name: variable.name,
              pattern,
              ...(printAttribute && {
                print: getFieldPrintPattern(printAttribute),
              }),
            }),
          ]
        }
      }

      return []
    }),
  }
}

function getEnumNode(
  declaration: EnumerationDeclaration,
  nodeNames: string[]
): EnumNodeDefinition {
  return {
    type: 'enum',
    name: declaration.name,
    pattern: getEnumParseAttribute(declaration)! as OrPattern,
    print: selfReferencePrintPattern(),
    fields: declaration.cases.flatMap((enumCase): Field[] => {
      const attributes = enumCase.data.attributes.map(
        attribute => new FunctionCallExpression(attribute)
      )
      const pattern = getParseAttribute(enumCase.data.name.name, attributes)
      const printAttribute = getPrintAttributeExpression(attributes)

      if (pattern) {
        return [
          field({
            name: enumCase.data.name.name,
            pattern,
            ...(printAttribute && {
              print: getFieldPrintPattern(printAttribute),
            }),
          }),
        ]
      }

      // Try to infer a pattern from the type annotation
      if (
        enumCase.data.associatedValues[0] &&
        enumCase.data.associatedValues[0].type === 'associatedValue'
      ) {
        const annotation = enumCase.data.associatedValues[0].data.annotation

        if (annotation && annotation.type === 'typeIdentifier') {
          const pattern = inferFieldPattern(
            enumCase.data.name.name,
            annotation,
            nodeNames
          )

          if (pattern) {
            return [
              field({
                name: enumCase.data.name.name,
                pattern,
                ...(printAttribute && {
                  print: getFieldPrintPattern(printAttribute),
                }),
              }),
            ]
          }
        }
      }

      return []
    }),
  }
}

export function buildParserDefinition(
  nodes: (EnumerationDeclaration | RecordDeclaration)[]
): Definition {
  const parserNodes = nodes.filter(isParser)
  const nodeNames = parserNodes.map(node => node.name)

  return {
    nodes: parserNodes.map(node => {
      if (node instanceof RecordDeclaration) {
        return getRecordNode(node, nodeNames)
      } else {
        return getEnumNode(node, nodeNames)
      }
    }),
  }
}

export function buildParser(
  nodes: (EnumerationDeclaration | RecordDeclaration)[]
): Parser {
  return new Parser(buildParserDefinition(nodes))
}
