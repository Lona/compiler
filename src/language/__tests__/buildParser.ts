import { LogicAST } from '@lona/serialization'
import { createFs } from 'buffs'
import fs from 'fs'
import path from 'path'
import { createHelpers } from '../../helpers'
import { createDeclarationNode } from '../../logic/nodes/createNode'
import { EnumerationDeclaration } from '../../logic/nodes/EnumerationDeclaration'
import { RecordDeclaration } from '../../logic/nodes/RecordDeclaration'
import { findNode, findNodes } from '../../logic/traversal'
import { buildLexer } from '../buildLexer'
import { buildParser, buildParserDefinition } from '../buildParser'
import { Printer } from '../Printer'

it('converts XML language', async () => {
  const source = createFs({
    'lona.json': JSON.stringify({}),
    'XMLLanguage.logic': fs.readFileSync(
      path.join(__dirname, './mocks/XMLLanguage.logic'),
      'utf8'
    ),
  })

  const helpers = createHelpers(source, '/')

  const logicFile = helpers.module.logicFiles.find(file =>
    file.sourcePath.endsWith('XMLLanguage.logic')
  )

  if (!logicFile) {
    throw new Error('Failed to find input file')
  }

  const mainEnum = findNode(
    logicFile.rootNode,
    (node): node is LogicAST.EnumerationDeclaration =>
      node.type === 'enumeration'
  )

  if (!mainEnum) {
    throw new Error('Failed to find tokens enum')
  }

  const lexer = buildLexer(new EnumerationDeclaration(mainEnum), helpers)

  const declarations = findNodes(
    logicFile.rootNode,
    node => node.type === 'enumeration' || node.type === 'record'
  ) as (LogicAST.EnumerationDeclaration | LogicAST.RecordDeclaration)[]

  const nodes = declarations.map(createDeclarationNode) as (
    | RecordDeclaration
    | EnumerationDeclaration
  )[]

  const definition = buildParserDefinition(nodes, helpers)

  expect(definition).toMatchSnapshot()

  const parser = buildParser(nodes, helpers)

  const tokens1 = lexer.tokenize(`<hello a="test" b='foo' />`)

  expect(parser.parse(tokens1, 'XMLElement')).toMatchSnapshot()

  const tokens2 = lexer.tokenize(`<OK>Some Text<Nested /></OK>`)

  const element2 = parser.parse(tokens2, 'XMLElement')

  expect(element2).toMatchSnapshot()

  const printer = new Printer(lexer.stateDefinitions, parser.definition)

  const formatted = printer.formatNode(element2, 'XMLElement')

  // console.log(formatted)

  // console.log(printer.print(formatted))
})
