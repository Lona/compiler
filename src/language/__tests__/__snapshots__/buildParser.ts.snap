// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts XML language 1`] = `
Object {
  "nodes": Array [
    Object {
      "fields": Array [
        Object {
          "name": "name",
          "pattern": Object {
            "type": "reference",
            "value": Object {
              "name": "name",
              "type": "token",
            },
          },
        },
        Object {
          "name": "value",
          "pattern": Object {
            "type": "reference",
            "value": Object {
              "name": "string",
              "type": "token",
            },
          },
        },
      ],
      "name": "XMLAttribute",
      "pattern": Object {
        "type": "sequence",
        "value": Array [
          Object {
            "type": "reference",
            "value": Object {
              "fieldName": "name",
              "nodeName": "XMLAttribute",
              "type": "field",
            },
          },
          Object {
            "type": "reference",
            "value": Object {
              "name": "equals",
              "type": "token",
            },
          },
          Object {
            "type": "reference",
            "value": Object {
              "fieldName": "value",
              "nodeName": "XMLAttribute",
              "type": "field",
            },
          },
        ],
      },
      "type": "record",
    },
    Object {
      "fields": Array [
        Object {
          "name": "charData",
          "pattern": Object {
            "type": "reference",
            "value": Object {
              "name": "text",
              "type": "token",
            },
          },
        },
        Object {
          "name": "comment",
          "pattern": Object {
            "type": "reference",
            "value": Object {
              "name": "comment",
              "type": "token",
            },
          },
        },
        Object {
          "name": "element",
          "pattern": Object {
            "type": "reference",
            "value": Object {
              "name": "XMLElement",
              "type": "node",
            },
          },
        },
      ],
      "name": "XMLElementContent",
      "pattern": Object {
        "type": "or",
        "value": Array [
          Object {
            "type": "reference",
            "value": Object {
              "fieldName": "charData",
              "nodeName": "XMLElementContent",
              "type": "field",
            },
          },
          Object {
            "type": "reference",
            "value": Object {
              "fieldName": "comment",
              "nodeName": "XMLElementContent",
              "type": "field",
            },
          },
          Object {
            "type": "reference",
            "value": Object {
              "fieldName": "element",
              "nodeName": "XMLElementContent",
              "type": "field",
            },
          },
        ],
      },
      "type": "enum",
    },
    Object {
      "fields": Array [
        Object {
          "name": "name",
          "pattern": Object {
            "type": "reference",
            "value": Object {
              "name": "open",
              "type": "token",
            },
          },
        },
        Object {
          "name": "attributes",
          "pattern": Object {
            "type": "many",
            "value": Object {
              "type": "reference",
              "value": Object {
                "name": "XMLAttribute",
                "type": "node",
              },
            },
          },
        },
        Object {
          "name": "content",
          "pattern": Object {
            "type": "many",
            "value": Object {
              "type": "reference",
              "value": Object {
                "name": "XMLElementContent",
                "type": "node",
              },
            },
          },
        },
      ],
      "name": "XMLElement",
      "pattern": Object {
        "type": "sequence",
        "value": Array [
          Object {
            "type": "reference",
            "value": Object {
              "fieldName": "name",
              "nodeName": "XMLElement",
              "type": "field",
            },
          },
          Object {
            "type": "reference",
            "value": Object {
              "fieldName": "attributes",
              "nodeName": "XMLElement",
              "type": "field",
            },
          },
          Object {
            "type": "or",
            "value": Array [
              Object {
                "type": "sequence",
                "value": Array [
                  Object {
                    "type": "reference",
                    "value": Object {
                      "name": "close",
                      "type": "token",
                    },
                  },
                  Object {
                    "type": "reference",
                    "value": Object {
                      "fieldName": "content",
                      "nodeName": "XMLElement",
                      "type": "field",
                    },
                  },
                  Object {
                    "type": "reference",
                    "value": Object {
                      "name": "slashOpen",
                      "type": "token",
                    },
                  },
                  Object {
                    "type": "reference",
                    "value": Object {
                      "name": "name",
                      "type": "token",
                    },
                  },
                  Object {
                    "type": "reference",
                    "value": Object {
                      "name": "close",
                      "type": "token",
                    },
                  },
                ],
              },
              Object {
                "type": "reference",
                "value": Object {
                  "name": "slashClose",
                  "type": "token",
                },
              },
            ],
          },
        ],
      },
      "type": "record",
    },
  ],
}
`;

exports[`converts XML language 2`] = `
Object {
  "tokens": Array [],
  "type": "success",
  "value": Object {
    "attributes": Array [
      Object {
        "name": "a",
        "value": "test",
      },
      Object {
        "name": "b",
        "value": "foo",
      },
    ],
    "name": "hello",
  },
}
`;

exports[`converts XML language 3`] = `
Object {
  "tokens": Array [],
  "type": "success",
  "value": Object {
    "attributes": Array [],
    "content": Array [
      Object {
        "type": "charData",
        "value": "Some Text",
      },
      Object {
        "type": "element",
        "value": Object {
          "attributes": Array [],
          "name": "Nested",
        },
      },
    ],
    "name": "OK",
  },
}
`;
